src/main.js

```diff
@@ -0,0 +1,208 @@
+        'element, i.e. "example.imgix.net".');
+    }
+
+    if (this.settings.includeLibraryParam) {
+      this.settings.libraryParam = "js-" + VERSION;
+    }
+
+    this.settings.urlPrefix = this.settings.useHTTPS ? 'https://' : 'http://';
+  }
+
+  buildURL(path = '', params = {}) {
+    const sanitizedPath = this._sanitizePath(path);
+
+    const queryParams = this._buildParams(params);
+    if (!!this.settings.secureURLToken) {
+      const securedParams = this._signParams(sanitizedPath, queryParams);
+      return this.settings.urlPrefix + this.settings.domain + sanitizedPath + securedParams;
+    }
+    return this.settings.urlPrefix + this.settings.domain + sanitizedPath + queryParams;
+  }
+
+  _buildParams(params = {}) {
+    const queryParams = [];
+    if (this.settings.libraryParam) {
+      queryParams.push('ixlib=' + this.settings.libraryParam)
+    }
+
+    for (const key in params) {
+      const val = params[key];
+      const encodedKey = encodeURIComponent(key);
+
+      let encodedVal;
+      if (key.substr(-2) === '64') {
+        encodedVal = Base64.encodeURI(val);
+      } else {
+        encodedVal = encodeURIComponent(val);
+      }
+      queryParams.push(encodedKey + "=" + encodedVal);
+    }
+
+    if (queryParams[0]) {
+      queryParams[0] = "?" + queryParams[0];
+    }
+    return queryParams.join('&');
+  }
+
+  _signParams(path, queryParams) {
+    const signatureBase = this.settings.secureURLToken + path + queryParams;
+    const signature = md5(signatureBase);
+
+    if (queryParams.length > 0) {
+      return queryParams = queryParams + "&s=" + signature;
+    } else {
+      return queryParams = "?s=" + signature;
+    }
+  }
+
+  _sanitizePath(path) {
+    // Strip leading slash first (we'll re-add after encoding)
+    let _path = path.replace(/^\//, '');
+
+    if (/^https?:\/\//.test(_path)) {
+      // Use de/encodeURIComponent to ensure *all* characters are handled,
+      // since it's being used as a path
+      _path = encodeURIComponent(_path);
+    } else {
+      // Use de/encodeURI if we think the path is just a path,
+      // so it leaves legal characters like '/' and '@' alone
+      _path = encodeURI(_path).replace(/[#?:]/g, encodeURIComponent);
+    }
+
+    return '/' + _path;
+  }
+
+  buildSrcSet(path, params, options) {
+    const _options = options || {};
+    const _params = params || {};
+    const width = _params.w;
+    const height = _params.h;
+    const aspectRatio = _params.ar;
+
+    if ((width) || (height && aspectRatio)) {
+      return this._buildDPRSrcSet(path, _params, _options);
+    } else {
+      return this._buildSrcSetPairs(path, _params, _options);
+    }
+  }
+
+  _buildSrcSetPairs(path, params, options) {
+    const srcsetOptions = validateAndDestructureOptions(options);
+    const srcset = [];
+    const customWidths = options.widths;
+    const widthTolerance = srcsetOptions[0], minWidth = srcsetOptions[1], maxWidth = srcsetOptions[2];
```

Comment:
L122 and L125 had to be written in a weird way to workaround the fact we couldn't use a destructuring assignment. I think it can now be safely rewritten:
```suggestion
    const [widthTolerance, minWidth, maxWidth]  = validateAndDestructureOptions(options);
    const srcset = [];
    const customWidths = options.widths;
```